### PR TITLE
Structured Streaming fixes

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
This change contains a couple of fixes I found when working on the Structured-Streaming end-to-end sample

1) Change-Feed pull model wasn't correctly wiring up retry policies
2) We didn't allow users to specify that null values are ok (even when in the schema inference sample set all records had non-null values in a certain property)
3) Fixed a bug in the calculation of the latest offset (which resulted in not fully draining change feed when new rows are added after starting the processing)

**NEXT**: I have a draft for the Structured Streaming end-to-end sample. Will send a separate PR for it tomorrow and then start working on automating execution of the end-to-end samples with the pipeline work Matias did.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 1]
```diff
@@ -387,6 +388,12 @@ private object CosmosSchemaInferenceConfig {
     parseFromStringFunction = include => include.toBoolean,
     helpMessage = "Whether schema inference should include the system properties in the schema")
 
+  private val inferSchemaForceNullableProperties = CosmosConfigEntry[Boolean](key = "spark.cosmos.read.inferSchemaForceNullableProperties",
+    mandatory = false,
+    defaultValue = Some(false),
+    parseFromStringFunction = include => include.toBoolean,
+    helpMessage = "Whether schema inference should enforce inferred properties to be nullable - even when no null-values are contained in the sample set")
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:395]: what's the use case for this?

**Review by moderakh — APPROVED** [review]
LGTM. thanks

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedScan.scala [hunk 2]
```diff
@@ -29,8 +29,10 @@ private case class ChangeFeedScan
     schema
   }
 
-  // TODO fabianm Implement
-  override def description(): String = super.description()
+  override def description(): String = {
+    // Ignore trailing $'s in the class names for Scala objects
+    this.getClass.getName.stripSuffix("$")
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedScan.scala:34]: the default is this already, if you don't implement:
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/Scan.java#L59-L61

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala (hunk 1 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:395]: Use-case is when a dataset contains a column that in the vast majority of cases is non-null (tripType in the NYC dataset for example). So when using schema inference the sample set might not have a single row with null in the property - in that case the schema inference by default will make the property non-nullable (that was one of the user study feedbacks). This config setting allows to relax that and make properties created by schema inference nullable by default.

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedScan.scala (hunk 2 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/ChangeFeedScan.scala]: Fixed this with the new context that the description is used to create the physical query plan (for example returned by the explain command)

**Pull request merged**